### PR TITLE
DNSZone module

### PR DESCRIPTION
**Pull request by seocam**
Added module to handle ipa `dnszone-add`, `dnszone-del`, `dnszone-mod`, `dnszone-enable` and `dnszone-disable`.

This PR also proposes a base class to implement our modules in ansbile-freeipa.

### CODE REVIEW

File: plugins/modules/ipadnszone.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Authors:
+#   Sergio Oliveira Campos <seocam@redhat.com>
+#
+# Copyright (C) 2019 Red Hat
```
Comment (rjeffman) [inline plugins/modules/ipadnszone.py]: 2020, happy new year! :-)

**Comment by seocam** [comment]
@t-woerner I left a commit with the workaround for the serial bug so you can see it clearly. After the review I will squash it.

File: playbooks/dnszone/dnszone-absent.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+---
+- name: Playbook to remove DNS zone
```
Comment (rjeffman) [inline playbooks/dnszone/dnszone-absent.yml]: It is better to ensure presence/absence, than add/remove.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 3]
```diff
@@ -309,3 +349,159 @@ def is_ipv6_addr(ipaddr):
...
+        pass
+
+    def api_command(self, command, name=None, args=None):
+        if args is None:
+            args = {}
+
+        if name is None:
+            return api_command_no_name(self, command, args)
+
+        return api_command(self, command, name, args)
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py]: Wouldn't it be better to reimplement the api_command* functions here using *args and **kwargs?
I mean, there are a few API commands that can take more the one positional parameter, and with variable arguments, this calls would also be contemplated.
As we spoke before, I wanted to test it, but could not do it so far.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 4]
```diff
@@ -220,25 +240,35 @@ def compare_args_ipa(module, args, ipa):
...
             try:
                 arg_set = set(arg)
                 ipa_arg_set = set(ipa_arg)
             except TypeError:
                 if arg != ipa_arg:
-                    # module.warn("%s != %s" % (repr(arg), repr(ipa_arg)))
+                    module.debug(
+                        base_debug_msg
+                        + "Different values: %s %s" % (arg, ipa_arg)
+                    )
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py]: Is this a debug message left behind or is it intentional? (I kinda like it.)

File: plugins/module_utils/ansible_freeipa_module.py [hunk 5]
```diff
@@ -220,25 +240,35 @@ def compare_args_ipa(module, args, ipa):
...
                     return False
             else:
                 if arg_set != ipa_arg_set:
-                    # module.warn("%s != %s" % (repr(arg), repr(ipa_arg)))
+                    module.debug(
+                        base_debug_msg
+                        + "Different set content: %s %s"
+                        % (arg_set, ipa_arg_set,)
+                    )
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py]: Also here... accidental or intentional?

File: plugins/module_utils/ansible_freeipa_module.py [hunk 6]
```diff
@@ -309,3 +349,159 @@ def is_ipv6_addr(ipaddr):
...
+
+    def require_ipa_attrs_change(self, command_args, ipa_attrs):
+        equal = compare_args_ipa(self, command_args, ipa_attrs)
+        return not equal
+
+    def plog(self, value):
+        self.log(pformat(value))
+
+    def pdebug(self, value):
+        self.debug(pformat(value))
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py]: It should also provide warn and error wrappers.

File: tests/dnszone/test_dnszone_mod.yml [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+      ipaadmin_password: SomeADMINpassword
+      name: testzone.local
+      allow_query:
+        - 1.1.1.1
+        - 2.2.2.2
+        - 3.3.3.3
+    register: result
+    failed_when: result.changed
+
+  - name: Remove allow query.
```
Comment (rjeffman) [inline tests/dnszone/test_dnszone_mod.yml]: Maybe "Ensure no query allowed." ?

File: tests/dnszone/test_dnszone_mod.yml [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    ipadnszone:
+      ipaadmin_password: SomeADMINpassword
+      name: testzone.local
+      allow_query: []
+    register: result
+    failed_when: not result.changed
+
+  - name: Remove allow query, again.
```
Comment (rjeffman) [inline tests/dnszone/test_dnszone_mod.yml]: Same as line 281.

File: plugins/module_utils/ansible_freeipa_module.py (hunk 4 above)
Comment (seocam) [reply plugins/module_utils/ansible_freeipa_module.py]: Yes it was left intentionally.

That brings something I'd like to ask you and @t-woerner about. We are forcing debug mode in all modules (or at least the modules I've checked). Is that really needed? Why did we decide to do that over only enabling debug when it's needed.

File: plugins/module_utils/ansible_freeipa_module.py (hunk 6 above)
Comment (seocam) [reply plugins/module_utils/ansible_freeipa_module.py]: I was in doubt even about keeping the "plog". The reason is that pformat will result the logs to be multi-line, which is nice for human readying but a little problematic for parsing. I haven't added to all logging levels to not encourage module developers to use pformat.

What do you think about removing plog and only keeping pdebug?

Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py]: That would be nice, as debug is (usually) geared towards humans, and the other levels for logs that might be parsed.

File: plugins/module_utils/ansible_freeipa_module.py (hunk 3 above)
Comment (seocam) [reply plugins/module_utils/ansible_freeipa_module.py]: I've added this commit to test that:

https://github.com/freeipa/ansible-freeipa/pull/224/commits/632e1e0c1842c1ef4a1fa5e4e0cb1cf39e831c6e

I'm not sure I like it. It works and I'm ok with the kwargs version I'm just not sure about replacing name by *args. It think it was a bit harder to understand and also seems to be masking the actual parameter.

You mentioned that a few commands can take more than one positional parameter, right? Could you point me to those so I can take a look and see if I can think of an better proposal?

Comment (seocam) [reply plugins/module_utils/ansible_freeipa_module.py]: After discussing on video call we decided to drop this idea for now.

File: plugins/modules/ipadnszone.py [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+    type: bool
+  skip_nameserver_check:
+    description: Force DNS zone creation even if nameserver is not resolvable
+    required: false
+    type: bool
+"""  # noqa: E501
+
+EXAMPLES = """
+---
+- ipadnszone:
```
Comment (rjeffman) [inline plugins/modules/ipadnszone.py]: For each of this example we usually add a comment `#` to explain what the example should do if used.

File: plugins/module_utils/ansible_freeipa_module.py [hunk 10]
```diff
@@ -88,6 +93,7 @@ def valid_creds(module, principal):
     if creds and \
        creds.lifetime > 0 and \
        "%s@" % principal in creds.name.display_as(creds.name.name_type):
+
```
Comment (rjeffman) [inline plugins/module_utils/ansible_freeipa_module.py]: Please, remove this empty line.

File: plugins/module_utils/ansible_freeipa_module.py (hunk 4 above)
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py]: I'm not sure why debug mode is always enable.

Comment (seocam) [reply plugins/module_utils/ansible_freeipa_module.py]: What about, removing that in the class? It wouldn't affect the modules not using the class.

File: README-dnszone.md [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+  - name: Ensure zone is present.
+    ipadnszone:
+      ipaadmin_password: SomeADMINpassword
+      name: testzone.local
+      state: present
+
+```
+
+
+Example playbook to create a DNS zone with all currently supported args:
```
[the remark this answers is not in the document]
Comment (seocam) [reply README-dnszone.md]: I'll change to be consistent with all modules we already have but those are actually module arguments and not variables (as you can see in https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#starting-a-new-module).

File: plugins/modules/README.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+* Code:
+  * A module file placed in `plugins/modules/<ipa_module_name>.py`
+
+* Documentation:
+  * `README-<module_name>.md` file in the root directory and linked from the main README.md
+  * Example playbooks in `playbooks/<module_name>/` directory
+
+* Tests:
+  * Test cases (also playbooks) defined in `tests/<module_name>/test_<something>.yml`. ore than one file is also accptable)
```
[the remark this answers is not in the document]
Comment (seocam) [reply plugins/modules/README.md]: Yeap. The line is a bit confusing as well. I'll update it.

File: plugins/module_utils/ansible_freeipa_module.py (hunk 4 above)
Comment (rjeffman) [reply plugins/module_utils/ansible_freeipa_module.py]: I'm not sure. But adding it later is easy, and we can add only if needed. Please, remove it.

Comment (seocam) [reply plugins/module_utils/ansible_freeipa_module.py]: Done.

**Pull request merged**